### PR TITLE
B feature oauth2

### PR DESCRIPTION
**Pull request by luinnarn**
Issue #16: basic Google login functionality. At this point, the only check is whether the signed in user is from the execom.eu domain.

### CODE REVIEW

File: src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java [hunk 1]
```diff
@@ -27,6 +28,18 @@ public void handleEntityExistsException(HttpServletRequest request, Exception ex
     logException(request, exception);
   }
 
+  @ResponseStatus(HttpStatus.CONFLICT) // 409
+  @ExceptionHandler(DataIntegrityViolationException.class)
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java]: Is this exception thrown only if a unique key constraint is violated, or there may be some other cases when this is also thrown (e.g., a relation between two tables is not satisfied)?

File: src/main/java/eu/execom/hawaii/api/controller/SignInController.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    if (!email.endsWith(EXECOM_DOMAIN)) {
+      return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+    }
+
+    User user = userService.getByEmail(email);
+
+    final Collection<GrantedAuthority> authorities = new ArrayList<>();
+    authorities.add(new SimpleGrantedAuthority(user.getUserRole().name()));
+
+    final Authentication authentication = new PreAuthenticatedAuthenticationToken(user, null, authorities);
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/api/controller/SignInController.java]: Please remove `final` keywords, we do not use it in our code base.

File: src/main/java/eu/execom/hawaii/service/UserService.java [hunk 3]
```diff
@@ -31,13 +32,26 @@ public UserService(UserRepository userRepository) {
...
   }
 
+  /**
+   * Retrieves a User with a specific id
+   *
+   * @param id User email
+   * @return User with specified email if exists
+   * @throws EntityNotFoundException if a user with given id is not found
+   */
+  @Transactional
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/service/UserService.java]: Please add `readOnly = true`.

File: src/main/java/eu/execom/hawaii/service/UserService.java [hunk 4]
```diff
@@ -31,13 +32,26 @@ public UserService(UserRepository userRepository) {
...
+  public User getUserById(Long id) {
+    return userRepository.getOne(id);
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/service/UserService.java]: Please add `readOnly = true`.

File: src/main/java/eu/execom/hawaii/service/UserService.java [hunk 5]
```diff
@@ -31,13 +32,26 @@ public UserService(UserRepository userRepository) {
...
+  }
+
   /**
    * Retrieves a User with a specific email.
    *
    * @param email User email
    * @return User with specified email if exists
    * @throws EntityNotFoundException if a user with given email is not found
    */
+  @Transactional
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/hawaii/service/UserService.java]: Please add `readOnly = true`.

File: src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java (hunk 1 above)
Comment (luinnarn) [reply src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java]: This is also thrown in other cases such as when a foreign key constraint or a NotNull constraint is not satisfied.

Comment (stanislavgrujic) [reply src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java]: Knowing that and reading this https://httpstatuses.com/409, does that behavior fit into description?

Comment (luinnarn) [reply src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java]: 409 should be thrown whenever a resource conflict would be caused by fulfilling the request. Doesn't a foreign key or not null constraint represent such a resource conflict?

Comment (stanislavgrujic) [reply src/main/java/eu/execom/hawaii/api/controller/GlobalExceptionHandler.java]: 4xx are client errors. Client cannot affect the foreign key, and null values should be validated upon receiving the client request, not on the database error.

**Pull request merged**